libs/nss.c

```diff
@@ -57,53 +57,80 @@ static FILE *tcb_safe_open(const char *file, const char *name)
 	return f;
 }
 
-int _nss_tcb_getspnam_r(const char *name, struct spwd *__result_buf,
-    char *__buffer, size_t __buflen, struct spwd **__result)
+enum nss_status _nss_tcb_getspnam_r(const char *name,
+    struct spwd *__result_buf, char *__buffer, size_t __buflen, int *__errnop)
 {
 	FILE *f;
 	char *file;
 	int retval, saved_errno;
+	struct spwd **result_buf_ptr = NULL;
 
-	if (asprintf(&file, TCB_FMT, name) < 0)
+	if (asprintf(&file, TCB_FMT, name) < 0) {
+		*__errnop = EAGAIN;
 		return NSS_STATUS_TRYAGAIN;
+	}
+
 	f = tcb_safe_open(file, name);
 	free(file);
-	if (!f)
+	if (!f) {
+		*__errnop = ENOENT;
 		return NSS_STATUS_UNAVAIL;
+	}
 
-	retval = fgetspent_r(f, __result_buf, __buffer, __buflen, __result);
+	retval = fgetspent_r(f, __result_buf, __buffer,
+	    __buflen, result_buf_ptr);
 	saved_errno = errno;
 	fclose(f);
 	errno = saved_errno;
-	if (!retval)
+	if (!retval) {
+		if (!*result_buf_ptr) {
+			*__errnop = ENOENT;
```

Comment:
Should we `*__errnop = errno` to use whatever `fgetspent_r` reported?

Reply:
Nope, as `fgetspent_r` by [documentation](https://man7.org/linux/man-pages/man3/fgetspent_r.3.html) indicates the error level by it return value; `buf_ptr == NULL` in this case means "no entry found".

By NSS specs that shall be indicated by `errnop == ENOENT` and `NSS_STATUS_NOTFOUND`.

We should on the other hand revise the follwing switch/case block to operate on the retval of `fgetspent_r`.

Reply:
Correction:  Using `saved_errno` for the switch/case is correct as I've misread the manpage for `fgetspent_r`.

Reply:
Another correction: I didn't misread, `fgetspent_r` really returns the error level: "The reentrant functions return zero on success.  In case of error, an error number is returned."